### ISSUE TITLE
Replication Groups

### ISSUE THREAD

**Issue by allen13** [VictoriaMetrics/VictoriaMetrics#541]
**Feature**
Add ability to selectively replicate to different groups of servers for multiple AZ HA setups.

**Solution**
```
--storageNode=group1/az1node1:8401
--storageNode=group1/az1node2:8401
--storageNode=group2/az2node1:8401
--storageNode=group2/az2node2:8401
```

Naive code sketch for 
https://github.com/VictoriaMetrics/VictoriaMetrics/blob/cluster/app/vminsert/netstorage/insert_ctx.go#L142
```go
for _,replicationGroup := range replicationGroups {
    if err := br.pushTo(replicationGroup.storageNodes[i]); err != nil && firstErr == nil {
        firstErr = err
    }
}
```

**Alternatives**

- Prometheus or vmagent multiple remote write endpoints. The queue buffers consume a lot of extra memory on prom where memory is already a challenge. This is probably better in vmagent but can't use that until it's compatable with prometheus operator.
- Something like [goreplay](https://github.com/buger/goreplay) the would duplicate the traffic. 

**Notes**

Implementing this functionality in vminsert is always going to be the optimal solution especially when prometheus is not the only source of data and the queue setup is not available.

**Comment by allen13** [comment]
I'm willing to attempt this myself but may need guidance on how it should be implemented. I realize making changes to the netstorage driver is non-trivial.

**Comment by valyala** [comment]
> I'm willing to attempt this myself but may need guidance on how it should be implemented. I realize making changes to the netstorage driver is non-trivial.

Just send a pull request with draft implementation and we'll see how it could be merged.

**Comment by allen13** [comment]
Initial attempt submitted. Ended up having to do a significant amount of refactoring to add replication groups and efficiently isolate the state between the two. Still needs more testing and review. I've successfully tested in influx insert handler. It was able to duplicate the insert between two different vmstorage nodes.

Is there a suite of integration tests that I could use to better vet the code?

**Comment by valyala** [comment]
@allen13 , thanks for the [pull request](https://github.com/VictoriaMetrics/VictoriaMetrics/pull/591)! It will be reviewed soon.

> Is there a suite of integration tests that I could use to better vet the code?

Unfortunately there are no integrations tests for cluster version of VictoriaMetrics at the moment. There are only [some integration tests for single-node VictoriaMetrics](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/victoria-metrics/main_test.go). It would be great if community could help creating integration tests for cluster version.

**Comment by valyala** [comment]
Replication groups can be implemented on top of [multi-level cluster setup](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#multi-level-cluster-setup). For example, the top level of `vminsert` nodes can be configured with `-replicationFactor=2` and with `-storageNode=vminsert-az1:8400,vminsert-az2:8400,vminsert-az3:8400` command-line flags, where `vminsert-azN` are `vminsert` nodes at the second level in each of availability zones (az1, az2, az3). Then `vminsert` nodes at the second level must run with `-clusternativeListenAddr=:8400` command-line flag in order to be able to accept data from top-level `vminsert` nodes and spread the accepted data among `vmstorage` nodes in the corresponding AZ. The functionality is available starting from the commit 4a5f45c77e5031c1de1fbe4b53244298c76c7e13 , which will be included in the next release (likely v1.60.0).

**Comment by valyala** [comment]
The ability to write data from one `vminsert` node to another `vminsert` node is available starting from [v1.60.0](https://github.com/VictoriaMetrics/VictoriaMetrics/releases). See [these docs](https://docs.victoriametrics.com/Cluster-VictoriaMetrics.html#multi-level-cluster-setup) for details. Closing the issue as done.
